libethereum/BlockChainSync.h

```diff
@@ -0,0 +1,213 @@
+*/
+/** @file BlockChainSync.h
+ * @author Gav Wood <i@gavwood.com>
+ * @date 2014
+ */
+
+#pragma once
+
+#include <mutex>
+
+#include <libdevcore/Guards.h>
+#include <libdevcore/RangeMask.h>
+#include <libethcore/Common.h>
+#include <libp2p/Common.h>
+#include "CommonNet.h"
+#include "DownloadMan.h"
+
+namespace dev
+{
+
+class RLPStream;
+
+namespace eth
+{
+
+class EthereumHost;
+class BlockQueue;
+class EthereumPeer;
+
+/**
+ * @brief Base BlockChain synchronization strategy class.
+ * Syncs to peers and keeps up to date. Base class handles blocks downloading but does not contain any details on state transfer logic.
+ */
+class BlockChainSync: public HasInvariants
+{
+public:
+	BlockChainSync(EthereumHost& _host);
+	virtual ~BlockChainSync();
+	void abortSync(); ///< Abort all sync activity
+
+	DownloadMan const& downloadMan() const;
+	DownloadMan& downloadMan();
+
+	/// @returns true is Sync is in progress
+	virtual bool isSyncing() const = 0;
+
+	/// Called by peer to report status
+	virtual void onPeerStatus(EthereumPeer* _peer);
+
+	/// Called by peer once it has new blocks during syn
+	virtual void onPeerBlocks(EthereumPeer* _peer, RLP const& _r);
+
+	/// Called by peer once it has new blocks
+	virtual void onPeerNewBlock(EthereumPeer* _peer, RLP const& _r);
+
+	/// Called by peer once it has new hashes
+	virtual void onPeerNewHashes(EthereumPeer* _peer, h256s const& _hashes) = 0;
+
+	/// Called by peer once it has another sequential block of hashes during sync
+	virtual void onPeerHashes(EthereumPeer* _peer, h256s const& _hashes) = 0;
+
+	/// Called by peer when it is disconnecting
+	virtual void onPeerAborting(EthereumPeer* _peer) = 0;
+
+	/// @returns Synchonization status
+	virtual SyncStatus status() const = 0;
+
+	static char const* stateName(SyncState _s) { return s_stateNames[static_cast<int>(_s)]; }
+
+protected:
+	//To be implemented in derived classes:
+	/// New valid peer appears
+	virtual void onNewPeer(EthereumPeer* _peer) = 0;
+
+	/// Peer done downloading blocks
+	virtual void peerDoneBlocks(EthereumPeer* _peer) = 0;
+
+	/// Resume downloading after witing state
+	virtual void continueSync() = 0;
+
+	/// Restart sync
+	virtual void restartSync() = 0;
+
+	/// Called after all blocks have been donloaded
+	virtual void completeSync() = 0;
+
+	/// Enter waiting state
+	virtual void pauseSync() = 0;
+
+	/// Restart sync for given peer
+	virtual void resetSyncFor(EthereumPeer* _peer, h256 _latestHash, u256 _td) = 0;
+
+	EthereumHost& host() { return m_host; }
+	EthereumHost const& host() const { return m_host; }
+
+	/// Estimates max number of hashes peers can give us.
+	unsigned estimateHashes();
+
+	/// Request blocks from peer if needed
+	void requestBlocks(EthereumPeer* _peer);
+
+private:
+	static char const* const s_stateNames[static_cast<int>(SyncState::Size)];
+	bool invariants() const override = 0;
+	EthereumHost& m_host;
+	HashDownloadMan m_hashMan;
+
+protected:
```

Comment:
customary to place class sections in increasing order of privacy: `public`, `protected`, then `private`.